functional/include/fun/option/option_inner.h

```diff
@@ -59,81 +60,105 @@ struct OptionUnion<T, std::enable_if_t<std::is_empty<T>::value>>: private T {
     return true;
   }
 
-  T unwrap() { return static_cast<T&&>(*this); }
+  T dump() {
+    _variant = Tag::NONE;
+    return static_cast<T&&>(*this);
+  }
 
   template <class... Args>
   void emplace(Args&&... args) {
-    static_cast<T&>(*this) = T(std::forward<Args>(args)...);
+    // Constructing a temporary and throwing it away mostly replicates the
+    // observable compile-time behavior of a normal emplace().
+    void(T(std::forward<Args>(args)...));
     _variant = Tag::SOME;
   }
-
 };
 
 //------------------------------------------------------------------------------
 template <class T>
-struct OptionUnion<T&> {
+class OptionUnion<T&> {
   using Self = OptionUnion<T&>;
 
   T* _ptr = nullptr;
-
+public:
   ~OptionUnion() = default;
 
   OptionUnion(const Self&) = default;
   Self& operator=(const Self&)  = default;
 
-  Self clone() const { return Self(*this); }
+  Self clone() const { return *this; }
+
+  OptionUnion(Self&& other) noexcept: _ptr(other._ptr) {
+    other._ptr = nullptr;
+  }
 
-  // For reference type move = copy
-  OptionUnion(Self&& other) = default;
-  Self& operator=(Self&&) = default;
+  Self& operator=(Self&& other) noexcept {
+    if (this != &other) {
+      _ptr = other._ptr;
+      other._ptr = nullptr;
+    }
+    return *this;
+  }
 
   OptionUnion() = default;
 
-  explicit OptionUnion(T& obj) : _ptr(&obj) {}
-  explicit OptionUnion(T* ptr) : _ptr(ptr) {}
+  explicit OptionUnion(T& obj) : _ptr(std::addressof(obj)) {}
 
-  OptionUnion(ForwardArgs, T& obj) : _ptr(&obj) {}
+  OptionUnion(ForwardArgs, T& obj) : _ptr(std::addressof(obj)) {}
 
   bool is_some() const { return _ptr ? true : false; }
 
   T* as_ptr() { return _ptr; }
 
   bool operator==(const Self& other) const { return _ptr == other._ptr; }
 
-  T& unwrap() { return *_ptr; }
+  T& dump() {
+    const auto ptr = _ptr;
+    _ptr = nullptr;
+    return *ptr;
+  }
 
   void emplace(T& ref) { _ptr = std::addressof(ref); }
 };
 
 //------------------------------------------------------------------------------
 template <class T, class En>
-struct OptionUnion {
+class OptionUnion {
   using Self = OptionUnion;
 
   enum class Tag: std::uint8_t { NONE, SOME };
   Tag _variant;
+
   union {
     Unit _empty = {};
-    std::remove_const_t<T> _val;
+    T _val;
   };
 
-  ~OptionUnion() {
-    if (_variant == Tag::SOME) { _val.~T(); }
+  void* val_addr() {
+    return static_cast<void*>(std::addressof(_val));
+  }
+
+  void erase() {
+    if (is_some()) {
+      _variant = Tag::NONE;
+      _val.~T();
+      _empty = {};
+    }
   }
+public:
+  ~OptionUnion() { erase(); }
 
   OptionUnion(const Self& other) : _variant(other._variant) {
-    if (_variant == Tag::SOME) { new (&_val) T(other._val); }
+    if (_variant == Tag::SOME) {
+      ::new (val_addr()) T(other._val);
```

Comment:
Did one of the build environments force you to put the `::` in front of `new`?

Reply:
To answer this (and the comment above as well):

Placement `new` syntax desugars to normal calls to `operator new()` with extra arguments. In the absence of any user-defined `operator new()` overloads, placement `new` works the way it does because the non-replaceable `operator new(std::size_t, void*)` overload just returns the pointer it's given, resulting in the new object being constructed at that location. It's possible, though, for someone to define their own `operator new(std::size_t, void*)` at class scope, and also possible for someone to define overloads that take non-`void` pointers. Prefixing with `::` and ensuring we only ever pass in `void*` makes sure we get the overload we were expecting.

Source:
https://en.cppreference.com/w/cpp/memory/new/operator_new

Reply:
In one of my personal projects, I wrote a `template <class T, class... Args> void construct_at(T*, Args&&...)` function that used this trick to guarantee it always had placement behavior. It might be worth adding something along these lines as a utility function; the current code layout is rather ugly, and returning `void*` from `val_addr()` is kinda a cludge to reduce how many `static_cast`s I had to write.

Reply:
Please replace with `construct_at`, I agree that seems better.